### ISSUE TITLE
IPv6 Compliance RFC4861: Redirected On-link: Valid (Hosts Only) [v6LC.2.3.1]

### ISSUE THREAD

**Issue by LiveFreeAndRoam** [systemd/systemd#31438]
### systemd version the issue has been seen with

255

### Used distribution

Ubuntu 23

### Linux kernel version used

6.5.0-17-generic

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

#### Test v6LC.2.3.1: Redirected On-link: Valid (Hosts Only)
**Purpose**: Verify that a host properly processes valid Redirect messages when redirected on-link.
**Reference**: RFC 4861, Section 4.6.1, 4.6.3, and 8.3

#### Network Configuration
```
       TR1
        |
    +---+---+
    |       |
   TN1     HUT
```

#### Procedure
| Step | Action | Expected Behavior of HUT<sup>(1)</sup> |
|--------|--------|--------|
| 1 | TR1 forwards `Echo request Src="TN1 Offlink GA"` to HUT GA<sup>(2)</sup>. | HUT sends `Echo reply` to TR1 as first hop. |
| 2 | TR1 sends `Redirect Dst="TN1 GA" Target="TN1 GA"` to HUT. | |
| 3 | TR1 forwards `Echo request Src="TN1 Offlink GA"` to HUT GA. | HUT sends `NS` for TN1's GA and `Echo reply` to TN1 on-link. |
* **(1)** - Host Under Test
* **(2)** - Global Address

The procedure is repeated with the Redirect options changing for each run.

| Run | TLLA | Redirected Packet Option |
|--------|--------|--------|
| A | No | No |
| B | No | Yes |
| C | Yes | No |
| D | Yes | Yes | 

### Unexpected behaviour you saw

`networkd` ignores the `Redirect` message.  In Step 3, `networkd` issues its `Ping reply` with first-hop being TR1.

### Steps to reproduce the problem

[v6LC_2_3_01a.zip](https://github.com/systemd/systemd/files/14366217/v6LC_2_3_01a.zip)

**Comment by yuwata** [comment]
> `networkd` ignores the `Redirect` message. In Step 3, `networkd` issues its `Ping reply` with first-hop being TR1.

Yes. Currently, sd-ndisc does not support Redirect message.

**Comment by LiveFreeAndRoam** [comment]
Is this something you can add or should I look at it myself?

**Comment by yuwata** [comment]
I will take a look.

**Comment by LiveFreeAndRoam** [comment]
Thank you @yuwata. I'll be ready to verify the algorithm as you make progress.

FYI, there are quite a few IPv6 Conformance tests related to the Redirect message.  I see we currently have 18 failing tests:

Test | Title
-- | --
v6LC_2_3_01a | Redirected   On-link: Valid (Hosts Only) (no TLLA or Redirect Pkt Option)
v6LC_2_3_01b | Redirected On-link:   Valid (Hosts Only) (no TLLA Option)
v6LC_2_3_01c | Redirected On-link:   Valid (Hosts Only) (no Redirect Pkt Option)
v6LC_2_3_01d | Redirected On-link:   Valid (Hosts Only) (TLLA and Redirect Pkt Option)
v6LC_2_3_02a | Redirected On-link:   Suspicious (Hosts Only) (Option unrecognized)
v6LC_2_3_02b | Redirected On-link:   Suspicious (Hosts Only) (Reserved field is non-zero)
v6LC_2_3_02c | Redirected On-link:   Suspicious (Hosts Only) (Target Address no Covered by On-link Prefix)
v6LC_2_3_04a | Redirected to   Alternate Router: Valid (Hosts Only) (No TLLA or Redirect Pkt Option)
v6LC_2_3_04b | Redirected to   Alternate Router: Valid (Hosts Only) (no TLLA Option)
v6LC_2_3_04c | Redirected to   Alternate Router: Valid (Hosts Only) (No Redirect Option)
v6LC_2_3_04d | Redirected to   Alternate Router: Valid (Hosts Only) (TLLA and Redirectd Packet Option)
v6LC_2_3_05a | Redirected to   Alternate Router: Valid (Hosts Only) (Option Unrecognized)
v6LC_2_3_05b | Redirected to   Alternate Router: Valid (Hosts Only) (Reserved field is non-zero)
v6LC_2_3_07 | Redirected Twice   (Hosts Only)
v6LC_2_3_08a | Invalid Option (Hosts   Only) (Path MTU Option)
v6LC_2_3_08b | Invalid Option (Hosts   Only) (Prefix Information Option)
v6LC_2_3_08c | Invalid Option (Hosts   Only) (Source Link-layer Address Option)
v6LC_2_3_09 | No Destination Cache   Entry (Hosts Only)

**Comment by yuwata** [comment]
Waiting in #30569.

**Comment by LiveFreeAndRoam** [comment]
I built against your [yuwata:network-next](https://github.com/yuwata/systemd/tree/network-next) branch.  Unfortunately, the tests are still failing.

I'll collect some more information and post it soon.

**Comment by LiveFreeAndRoam** [comment]
@yuwata, just to coordinate, is this ready for testing? The test still fails with your last two commits, but just want to check whether I should do the analysis or if you're not ready for me to re-run the tests.  Thanks.

**Comment by yuwata** [comment]
Sorry for wasting your time, I guess that's not ready yet.

**Comment by yuwata** [comment]
But, if you already have some info, please let me know.

**Comment by LiveFreeAndRoam** [comment]
No worries.  I now have a streamlined test environment, so I can pretty quickly run through the tests.  If they pass, it's easy. If they fail, it may take me the better part of an hour (sometimes more) to do the research and analysis and report my findings.

I'm happy to run the tests anytime you want something checked.  And once you feel it is in good shape, I'll gladly dive into the failing cases and create a report.

I still have other failing IPv6 tests cases to chase.  So I'll stay on track with those, but let me know when you would like to have deeper analysis on this case and I'll dig in.

**Comment by LiveFreeAndRoam** [comment]
@yuwata,

I have some new information. Let me know if there is more I can do here.

The reason the tests fail is because `networkd` permanently installs a redirect route that is not being cleaned up after the test.  This causes subsequent tests to fail.

My observation is that `networkd` handles redirects differently to the `native kernel` processing. When `networkd` processes the redirect, we see the following route permanently created:

```
2001:2:0:1001:200:10ff:fe10:1180 dev enp0s9 proto ra metric 1024 mtu 1500 hoplimit 64 pref mediu
```
Q: Regarding the `proto`, it is set to `proto ra`, should this be `proto redirect`?

Compared to when the `native kernel` processes the `RA` and `Redirect`, this route is not created when the `Redirect` packet is processed.  In fact, I have not been able to observe how the kernel is maintaining its redirect route, other than to say it works, whatever it is doing. The kernel cleans up its redirect after each test.

As an encouraging side note: With `networkd`, I can get the tests to pass, as long as I manually delete the permanent redirect route before running each test.

I have attached a zip of the PCAP, containing just 3 packets that may help you investigate further. You can use the following command to send each packet one-at-a-time:

```
$ sudo tcpreplay -o -i enp0s9 tcpreplay-redirect.pcapng 
```

[tcpreplay-redirect.zip](https://github.com/systemd/systemd/files/14428576/tcpreplay-redirect.zip)

Here's a summary of the three packets:

1. RA - to setup global prefix with long lifetime
2. Redirect - for the redirect
3. RA - to clear the global prefix by setting lifetime to zero.

Here is the `networkd` debug log...

```
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 0
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Removing NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Found an existing route that conflicts with new route based on a received RA, removing.
enp0s9: Removing NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (configured,removing): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Setting SLAAC addresses and router.
enp0s9: Refuse to configure NDisc route with zero lifetime.
enp0s9: Forgetting removed NDisc route (n/a): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1060: Router=yes, Solicited=no, Override=yes
enp0s9: NDISC: Invoking callback for 'neighbor' event.
enp0s9: Forgetting removed NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
```

Even though the prefix has expired, I still see an IP address persists and the redirected route also persists.

```
$ ip -6 r show dev enp0s9
2001:2:0:1001:200:10ff:fe10:1180 proto ra metric 1024 mtu 1500 hoplimit 64 pref medium
fe80::/64 proto kernel metric 256 pref medium

$ ip -6 a show dev enp0s9
4: enp0s9: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    inet6 2001:2:0:1000:a00:27ff:fe1a:432/64 scope global dynamic mngtmpaddr noprefixroute 
       valid_lft 64670sec preferred_lft 64670sec
    inet6 fe80::a00:27ff:fe1a:432/64 scope link 
       valid_lft forever preferred_lft forever
```

The route that remains

**Comment by LiveFreeAndRoam** [comment]
@yuwata, this might be helpful...

Using the three packets in the previous post, this highlights the difference between `native kernel` and `networkd`.

@yuwata, this might be helpful...

Using the three packets in the previous post, this highlights the difference in behaviour between `native kernel` and `networkd`. (As an aside, I still haven't found a way to show the redirect route when using the _native kernel_.)

| Packet| native-kernel | networkd |
|--------|--------|--------|
| **0. Initial state.**<br />Same for both. | ```$ ip -6 r show dev enp0s9```<br />```fe80::/64 proto kernel metric 256 pref medium``` <br /><br />```$ ip -6 r get 2001:2:0:1001:200:10ff:fe10:1180```<br />```RTNETLINK answers: Network is unreachable``` |  ```$ ip -6 r show dev enp0s9```<br />```fe80::/64 proto kernel metric 256 pref medium``` <br /><br />```$ ip -6 r get 2001:2:0:1001:200:10ff:fe10:1180```<br />```RTNETLINK answers: Network is unreachable``` |
| **1. RA prefix=2001:2:0:1000::/64 life=65535**<br /><br />**Note:** Both _networkd_ and _native kernel_ behave the same (for the relevant parts). | ```$ ip -6 r show dev enp0s9```<br />```2001:2:0:1000::/64 proto kernel metric 256 expires 65530sec pref medium```<br />```fe80::/64 proto kernel metric 256 pref medium```<br />```default via fe80::200:10ff:fe10:1060 proto ra metric 1024 expires 8995sec hoplimit 64 pref medium```<br /><br />```$ ip -6 r get 2001:2:0:1001:200:10ff:fe10:1180```<br />```2001:2:0:1001:200:10ff:fe10:1180 from :: via fe80::200:10ff:fe10:1060 dev enp0s9 proto ra src 2001:2:0:1000:a00:27ff:fe1a:432 metric 1024 hoplimit 64 pref medium``` | ```ip -6 r show dev enp0s9```<br />```2001:2:0:1000::/64 proto ra metric 1024 expires 65526sec mtu 1500 hoplimit 64 pref medium```<br />```fe80::/64 proto kernel metric 256 pref medium```<br />```default via fe80::200:10ff:fe10:1060 proto ra metric 1024 expires 8991sec mtu 1500 hoplimit 64 pref medium```<br /><br />```$ ip -6 r get 2001:2:0:1001:200:10ff:fe10:1180```<br />```2001:2:0:1001:200:10ff:fe10:1180 from :: via fe80::200:10ff:fe10:1060 dev enp0s9 proto ra src 2001:2:0:1000:a00:27ff:fe1a:432 metric 1024 mtu 1500 hoplimit 64 pref medium``` |
 | **2. Redirect to 2001:2:0:1001:200:10ff:fe10:1180**<br /><br />**Note:** _networkd_ creates redirect route. Also it use `proto ra` instead of `proto redirect`<br /><br />Where is the _native kernel_ route? | ```$ ip -6 r show dev enp0s9```<br />```2001:2:0:1000::/64 proto kernel metric 256 expires 65514sec pref medium```<br />```fe80::/64 proto kernel metric 256 pref medium```<br />```default via fe80::200:10ff:fe10:1060 proto ra metric 1024 expires 8979sec hoplimit 64 pref medium```<br /><br />```$ ip -6 r get 2001:2:0:1001:200:10ff:fe10:1180```<br />```2001:2:0:1001:200:10ff:fe10:1180 from :: dev enp0s9 src 2001:2:0:1000:a00:27ff:fe1a:432 metric 1024 hoplimit 64 pref medium``` | ```$ ip -6 r show dev enp0s9```<br />```2001:2:0:1000::/64 proto ra metric 1024 expires 64977sec mtu 1500 hoplimit 64 pref medium```<br />```2001:2:0:1001:200:10ff:fe10:1180 proto ra metric 1024 mtu 1500 hoplimit 64 pref medium```<br />```fe80::/64 proto kernel metric 256 pref medium```<br />```default via fe80::200:10ff:fe10:1060 proto ra metric 1024 expires 8442sec mtu 1500 hoplimit 64 pref medium```<br /><br />```$ ip -6 r get 2001:2:0:1001:200:10ff:fe10:1180```<br />```2001:2:0:1001:200:10ff:fe10:1180 from :: dev enp0s9 proto ra src 2001:2:0:1000:a00:27ff:fe1a:432 metric 1024 mtu 1500 hoplimit 64 pref medium```<br /> |
| **3. RA prefix=2001:2:0:1000::/64 life=0**<br /><br />**Note:** _networkd_ retains the redirect route, meanwhile the `native kernel` does not. | ```$ ip -6 r show dev enp0s9```<br />```fe80::/64 proto kernel metric 256 pref medium```<br /><br />```$ ip -6 r get 2001:2:0:1001:200:10ff:fe10:1180```<br />```RTNETLINK answers: Network is unreachable``` | ```$ ip -6 r show dev enp0s9```<br />```2001:2:0:1001:200:10ff:fe10:1180 proto ra metric 1024 mtu 1500 hoplimit 64 pref medium```<br />```fe80::/64 proto kernel metric 256 pref medium```<br /><br />```$ ip -6 r get 2001:2:0:1001:200:10ff:fe10:1180```<br />```2001:2:0:1001:200:10ff:fe10:1180 from :: dev enp0s9 proto ra src 2001:2:0:1000:a00:27ff:fe1a:432 metric 1024 mtu 1500 hoplimit 64 pref medium```<br /> |

**Comment by yuwata** [comment]
@LiveFreeAndRoam Thanks. That's much helpful. Could you run `ip -6 route` command with `table all`? If I understand correctly, the kernel configures a route based on Redirect.

**Comment by yuwata** [comment]
Or, there may be `ip rule`??

**Comment by yuwata** [comment]
> Q: Regarding the `proto`, it is set to `proto ra`, should this be `proto redirect`?

Nice catch!

**Comment by LiveFreeAndRoam** [comment]
> Could you run ip -6 route command with table all? If I understand correctly, the kernel configures a route based on Redirect.

Glad to.  I've done two runs with the same sequence as above: Initial state, RA setup prefix, Redirect, RA cleanup.

#### Native Kernel Behaviour
```
$ # Initial state
$ ip -6 r show table all dev enp0s9
fe80::/64 proto kernel metric 256 pref medium
local 2001:2:0:1000:a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
local fe80::a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
multicast ff00::/8 table local proto kernel metric 256 pref medium
 
$ # RA setup prefix
$ ip -6 r show table all dev enp0s9
2001:2:0:1000::/64 proto kernel metric 256 expires 65531sec pref medium
fe80::/64 proto kernel metric 256 pref medium
default via fe80::200:10ff:fe10:1060 proto ra metric 1024 expires 8996sec hoplimit 64 pref medium
local 2001:2:0:1000:a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
local fe80::a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
multicast ff00::/8 table local proto kernel metric 256 pref medium

$ # Redirect
$ ip -6 r show table all dev enp0s9
2001:2:0:1000::/64 proto kernel metric 256 expires 65515sec pref medium
fe80::/64 proto kernel metric 256 pref medium
default via fe80::200:10ff:fe10:1060 proto ra metric 1024 expires 8980sec hoplimit 64 pref medium
local 2001:2:0:1000:a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
local fe80::a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
multicast ff00::/8 table local proto kernel metric 256 pref medium

$ # Cleanup
$ ip -6 r show table all dev enp0s9
fe80::/64 proto kernel metric 256 pref medium
local 2001:2:0:1000:a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
local fe80::a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
multicast ff00::/8 table local proto kernel metric 256 pref medium
```

#### Networkd (network-next: 4c855b3d04 ) Behaviour
```
# Initial state
$ ip -6 r show table all dev enp0s9
fe80::/64 proto kernel metric 256 pref medium
local fe80::a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
multicast ff00::/8 table local proto kernel metric 256 pref medium

# RA setup prefix
$ ip -6 r show table all dev enp0s9
2001:2:0:1000::/64 proto ra metric 1024 expires 65517sec mtu 1500 hoplimit 64 pref medium
fe80::/64 proto kernel metric 256 pref medium
default via fe80::200:10ff:fe10:1060 proto ra metric 1024 expires 8982sec mtu 1500 hoplimit 64 pref medium
local 2001:2:0:1000:a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
local fe80::a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
multicast ff00::/8 table local proto kernel metric 256 pref medium

# Redirect
$ ip -6 r show table all dev enp0s9
2001:2:0:1000::/64 proto ra metric 1024 expires 65503sec mtu 1500 hoplimit 64 pref medium
2001:2:0:1001:200:10ff:fe10:1180 proto ra metric 1024 mtu 1500 hoplimit 64 pref medium
fe80::/64 proto kernel metric 256 pref medium
default via fe80::200:10ff:fe10:1060 proto ra metric 1024 expires 8968sec mtu 1500 hoplimit 64 pref medium
local 2001:2:0:1000:a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
local fe80::a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
multicast ff00::/8 table local proto kernel metric 256 pref medium

# RA Cleanup
$ ip -6 r show table all dev enp0s9
2001:2:0:1001:200:10ff:fe10:1180 proto ra metric 1024 mtu 1500 hoplimit 64 pref medium
fe80::/64 proto kernel metric 256 pref medium
local 2001:2:0:1000:a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
local fe80::a00:27ff:fe1a:432 table local proto kernel metric 0 pref medium
multicast ff00::/8 table local proto kernel metric 256 pref medium
```

PS: As an aside, I noticed the top-level `Makefile` file has been deleted. Was that on purpose? My VSCode IDE relies on it to build.  I'm sure I can figure out how to live without it, but it's just one more thing I have to chase.

**Comment by LiveFreeAndRoam** [comment]
> Or, there may be ip rule??

Good thought, but alas no dice.  I tested it with the `kernel native` configuration, and the output from `ip rule` didn't change throughout the test.

```
# Initial state
$ ip rule
0:	from all lookup local
32766:	from all lookup main
32767:	from all lookup default

# RA Setup prefix
$ ip rule
0:	from all lookup local
32766:	from all lookup main
32767:	from all lookup default
 
# Redirect
$ ip rule
0:	from all lookup local
32766:	from all lookup main
32767:	from all lookup default

# RA Cleanup
$ ip rule
0:	from all lookup local
32766:	from all lookup main
32767:	from all lookup default
```

**Comment by yuwata** [comment]
It seems you show IPv4 routing policy rule. Please use `ip -6 rule`.

**Comment by LiveFreeAndRoam** [comment]
Oh, shoot.  Sorry about that.  It's a similar result...

```
# Initial state
matt@ipv6logo-nut:~$ ip -6 rule
0:	from all lookup local
32766:	from all lookup main

# RA Setup prefix
$ ip -6 rule
0:	from all lookup local
32766:	from all lookup main
 
# Redirect
$ ip -6 rule
0:	from all lookup local
32766:	from all lookup main

# RA Cleanup
$ ip -6 rule
0:	from all lookup local
32766:	from all lookup main
```

**Comment by yuwata** [comment]
Hm. Interesting. But, in one of the above comment, the results of `ip route get` are mostly equivalent, except for `proto` and `mtu`. So, the behavior should be fine I guess, no?
In the revised version I pushed a half hour ago (eff03473a6b0ef7e637c00907b5022670501e4ff), Now the route for the redirected hosts are removed when a RA with lifetime=0 is received. Also proto should be now redirect.
Please try again.

**Comment by LiveFreeAndRoam** [comment]
Sounds good. I'll check it again on Monday.

**Comment by LiveFreeAndRoam** [comment]
With your latest changes, especially removal of the redirect-route when its prefix expires, 69 out of the 70 Redirect-tests are passing.

The one failing test is `v6LC.2.3.7 Redirected Twice`.

I'll write-up the details of the failing test in the next comment.  It will take me a while to collect all the info.

**Comment by LiveFreeAndRoam** [comment]
#### v6LC.2.3.7: Redirected Twice (Hosts Only)
**Purpose**: Verify that a host properly processes valid Redirect messages twice for the same 
destination.
**Reference**: [RFC4861 - Sections 8.3](https://datatracker.ietf.org/doc/html/rfc4861#section-8.3)

#### Network Configuration
```
   TR1 TR2  TR3
    |   |   |
    +---+---+
    |       |
   TN1     HUT
```
- **TR** = Test Router
- **TN** = Test Node
- **HUT** = Host Under Test

#### Procedure
| Step | Action | Expected Behaviour of HUT |
|--------|--------|--------|
| 1 | TR1 forwards an Echo Request from TN1 to the HUT. The Source Address is the off-link global address of TN1. The Destination Address is the GA<sup>(1)</sup> of the HUT. | The HUT should respond to the Echo Request using TR1 as a first hop, as it is the only router in the HUT’s Default Router List. |
| 2 | TR1 sends `Redirect Dst="TN1 GA" Target="TR2 LLA"` to HUT. | |
| 3 | TR1 sends `Redirect Dst="TN1 GA" Target=TR2 LLA"` to HUT. | |
| 4 | TR1 forwards an Echo Request from TN1 to the HUT. The Source Address is the off-link global address of TN1. The Destination Address is the global address of the HUT. | The HUT should respond to the Echo Request using TR2 as a first hop, indicating the HUT processed the Redirect message. |
| 5 |  TR2 sends `Redirect Dst="TN1 GA" Target="TR3 LLA"` to HUT. | |
| 6 |  TR1 forwards an Echo Request from TN1 to the HUT. The Source Address is the off-link global address of TN1. The Destination Address is the global address of the HUT. | The HUT should respond to the Echo Request using TR3 as a first hop, indicating the HUT processed the Redirect message. |

* **(1)** - Global Address

### Unexpected behaviour you saw

In Step 6, the HUT responded to the Echo request using TR2, when it should have used TR3.

### Steps to reproduce the problem
[v6LC_2_3_07-Net0-20240304123454.zip](https://github.com/systemd/systemd/files/14476120/v6LC_2_3_07-Net0-20240304123454.zip)


### Debug Log
```
#
# Preamble, setting everything up (ignore the preamble and jump to the next comment)
#
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 2h 30min
enp0s9: NDISC: Invoking callback for 'router' event.
Setting '/proc/sys/net/ipv6/neigh/enp0s9/base_reachable_time_ms' to '30000'
No change in value '30000', suppressing write
Setting '/proc/sys/net/ipv6/neigh/enp0s9/retrans_time_ms' to '1000'
No change in value '1000', suppressing write
Setting '/proc/sys/net/ipv6/conf/enp0s9/hop_limit' to '64'
No change in value '64', suppressing write
enp0s9: Requesting NDisc route (n/a): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc address (configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: Setting SLAAC addresses and router.
enp0s9: State changed: configured -> configuring
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: Received new foreign route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc route (requesting): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc route (requesting): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc address (requesting,configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: Received new NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Received updated NDisc address (configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.
enp0s9: Received updated NDisc address (configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: SLAAC addresses and routes set.
enp0s9: link_check_ready(): IPv4LL:no DHCPv4:no DHCPv6:no DHCP-PD:no NDisc:yes
enp0s9: State changed: configuring -> configured
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1060: Router=yes, Solicited=yes, Override=no
enp0s9: NDISC: Invoking callback for 'neighbor' event.
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1061: Router=yes, Solicited=yes, Override=no
enp0s9: NDISC: Invoking callback for 'neighbor' event.
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1062: Router=yes, Solicited=yes, Override=no
enp0s9: NDISC: Invoking callback for 'neighbor' event.
#
# First Redirect results in new route.
#
enp0s9: NDISC: Received Redirect message from fe80::200:10ff:fe10:1060: Target=fe80::200:10ff:fe10:1061, Destination=2001:2:0:1001:200:10ff:fe10:1180
enp0s9: NDISC: Invoking callback for 'redirect' event.
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:1001:200:10ff:fe10:1180/128, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: redirect, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc route (requesting): dst: 2001:2:0:1001:200:10ff:fe10:1180/128, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: redirect, scope: global, type: unicast, flags: n/a
enp0s9: Received new NDisc route (configured): dst: 2001:2:0:1001:200:10ff:fe10:1180/128, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: redirect, scope: global, type: unicast, flags: n/a
enp0s9: SLAAC addresses and routes set.

#
# This is the 2nd redirect. No route is configured for this 2nd Redirect.
#
enp0s9: NDISC: Received Redirect message from fe80::200:10ff:fe10:1061: Target=fe80::200:10ff:fe10:1062, Destination=2001:2:0:1001:200:10ff:fe10:1180
enp0s9: NDISC: Invoking callback for 'redirect' event.
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 0
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Removing NDisc route (configured): dst: 2001:2:0:1001:200:10ff:fe10:1180/128, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: redirect, scope: global, type: unicast, flags: n/a
enp0s9: Removing foreign route (configuring,configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Found an existing route that conflicts with new route based on a received RA, removing.
enp0s9: Removing NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (configured,removing): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Setting SLAAC addresses and router.
enp0s9: Forgetting removed foreign route (configuring): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Refuse to configure NDisc route with zero lifetime.
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1060: Router=yes, Solicited=no, Override=yes
enp0s9: NDISC: Invoking callback for 'neighbor' event.
enp0s9: Forgetting removed NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 0
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Setting SLAAC addresses and router.
enp0s9: State changed: configured -> configuring
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: Forgetting removed NDisc route (n/a): dst: 2001:2:0:1001:200:10ff:fe10:1180/128, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: redirect, scope: global, type: unicast, flags: n/a
enp0s9: Refuse to configure NDisc route with zero lifetime.
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1061: Router=yes, Solicited=no, Override=yes
enp0s9: NDISC: Invoking callback for 'neighbor' event.
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 0
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Setting SLAAC addresses and router.
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
Could not drop route, ignoring: No such process
enp0s9: Refuse to configure NDisc route with zero lifetime.
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1062: Router=yes, Solicited=no, Override=yes
enp0s9: NDISC: Invoking callback for 'neighbor' event.
Could not drop route, ignoring: No such process
```

**Comment by yuwata** [comment]
@LiveFreeAndRoam Could you test PR #30569 (https://github.com/yuwata/systemd/tree/network-next branch) again? The last issue should be fixed now. The code to support Redirect message is changed a lot, so, please test also already passed test cases. Thanks.

**Comment by LiveFreeAndRoam** [comment]
I ran through the 70 tests related to Redirects.  The only failing test is still **2.3.7 Redirect Twice**.  I wrote that up in the [comments above](https://github.com/systemd/systemd/issues/31438#issuecomment-1975636499).  Do you think I should write it up as a separate issue for tracking?

I'm about to be out of office for the next week, so now is a good time to reflect on our progress...

We initially had103 tests that failed, from 416 tests.  We are now down to just 3 tests that must pass, and 3 more which are desirable, but also optional.

In the relatively short time you've been resolving these, this is an amazing result @yuwata! 👏

The remaining failing tests are:

Group | Test # | Optional | Test Name | Status
-- | -- | -- | -- | --
Base | v6LC_2_2_23c | Optional | Processing   Router Advertisement with Route Information Option (Host Only) (Reserved   pref) | [FAIL](https://github.com/systemd/systemd/issues/28449)
Base | v6LC_2_2_23j | Optional | Processing Router   Advertisement with Route Information Option (Host Only) (Len 0, life 0) | [FAIL](https://github.com/systemd/systemd/issues/28449)
Ping | v6LC_2_2_25f | Optional | Processing Router   Advertisement DNS (Host Only) (Search List Option expired) | [FAIL](https://github.com/systemd/systemd/issues/31624)
Base | v6LC_2_3_07 |  Required | Redirected Twice   (Hosts Only) | [FAIL](https://github.com/systemd/systemd/issues/31438#issuecomment-1975636499)
DAD | v6LC_3_2_05c |  Required | Prefix-Information   Option Processing, Lifetime (Hosts Only) (life < remaining && life   < 2h) | [FAIL](https://github.com/systemd/systemd/issues/31449)
DAD | v6LC_3_2_07b |  Required | Resolving DAD   Conflicts (Host Only) (GA) | [FAIL](https://github.com/systemd/systemd/issues/31495)

Looking forward to more to come.

**Comment by LiveFreeAndRoam** [comment]
To simplify tracking, I have created a new issue for tracking v6LC.2.3.7.  See https://github.com/systemd/systemd/issues/31981.

Furhter, the issue in this thread (v6LC.2.3.1) is resolved and so this issue may be closed.

The v6LC2.3.1a,b,c,d all pass using:

```
$ git log -n 1
commit 59b72d6f47160c74a1d7f44bc3e2373b4e14dd58 (HEAD -> network-next, origin/network-next)
```

**Comment by yuwata** [comment]
Thank you! I will open a PR with relevant commits in `network-next` branch after #31373 is merged.

**Comment by yuwata** [comment]
-> #32104.
